src/QMCWaveFunctions/Jastrow/codegen/user_jastrow.py

```diff
@@ -0,0 +1,515 @@
+      {
+        std::string id_in;
+        std::string p_name;
+        OhmmsAttributeSet rAttrib;
+        rAttrib.add(id_in, "id");
+        rAttrib.add(p_name, "name");
+        rAttrib.put(cur);
+        %(read_param)s
+      }
+      cur = cur->next;
+    }
+    reset();
+    myVars.clear();
+    %(set_opt)s
+    return true;
+  }
+"""
+
+  read_param_str = """
+        if (p_name == "%(var_name)s")
+        {
+          %(id_var_name)s = id_in;
+          putContent(%(var_name)s, cur);
+          %(opt_var_name)s = %(do_opt)s;
+        }
+  """
+
+  set_opt_str = """
+    if (%(opt_var_name)s)
+      myVars.insert(%(id_var_name)s, %(var_name)s, %(opt_var_name)s, optimize::OTHER_P);
+  """
+
+  read_param = ""
+  set_opt = ""
+  for var_name in param_list:
+    repl = get_var_names(var_name)
+    repl['do_opt'] = 'true'
+    read_param += read_param_str%repl
+    set_opt += set_opt_str%repl
+
+  if input_param_list:
+    for var_name in input_param_list:
+      repl = get_var_names(var_name)
+      repl['do_opt'] = 'false'
+      read_param += read_param_str%repl
+
+
+  s = out_str%{"read_param":read_param, "set_opt":set_opt}
+
+  print('Creating XML input function')
+  print(s)
+  print('')
+
+  return s
+
+
+# Code for resetting the parameters
+def gen_reset_parameters(param_list):
+  out_str = """
+  void resetParameters(const opt_variables_type& active)
+  {
+    if (myVars.size())
+    {
+      int ia = myVars.where(0);
+      if (ia > -1)
+      {
+        int i = 0;
+        %(set_param)s
+      }
+      reset();
+    }
+  }
+  """
+
+  set_param = """
+        if (%(opt_var_name)s)
+          %(var_name)s = myVars[i++] = active[ia++];
+  """
+
+  set_param_str = ""
+  for var_name in param_list:
+    repl = get_var_names(var_name)
+    set_param_str += set_param%repl
+
+  s = out_str%{"set_param":set_param_str}
+
+  print('Creating reset parameters function')
+  print(s)
+  print('')
+
+  return s
+
+
+# Example XML input
+def gen_example_xml_input(param_list, input_param_list=None, cusp_param=None):
+  out_str = """
+   <jastrow name="Jee" type="Two-Body" function="user">
+      <correlation speciesA="u" speciesB="d">
+%(var_entries)s
+      </correlation>
+    </jastrow>
+   """
+  one_entry_str = """        <var id="j_%(var_name)s" name="%(var_name)s">1.0</var>\n"""
+
+  entries = ""
+
+  for var_name in param_list:
+    if var_name != cusp_param:
+      entries += one_entry_str%{'var_name':var_name}
+
+
+  if input_param_list:
+    for var_name in input_param_list:
+      entries += one_entry_str%{'var_name':var_name}
+
+  s = out_str%{'var_entries':entries}
+
+  print("Example XML Input section")
+  print(s)
+  print()
+
+
+
+dire_codegen_text = """
+/*
+ DO NOT MAKE PERMANENT EDITS IN THIS FILE
+ This file is generated from src/QMCWavefunctions/Jastrow/codegen/user_jastrow.py and UserFunctor.h.in
+
+ To make changes, edit UserFunctor.h.in and rerun user_jastrow.py.
+*/
+"""
+
+# A simple template replacement engine.
```

Comment:
Python's std library has a template mechanism and its syntax is very much like shell ${substitution} which is a plus.
https://www.python.org/dev/peps/pep-0292